### ISSUE TITLE
Atomic multi create-branch / create-commit / put-file

### ISSUE THREAD

**Issue by jdoliner** [pachyderm/pachyderm#3267]
**What is the goal / desired outcome?**

To be able to trigger pipelines atomically on inputs from multiple repos.

**If there is a way to accomplish this today via workaround, what does that require?**
This is not presently possible.

**(Optional) What is your proposal for a feature to solve this?**
All commands that can trigger pipelines (those which call `propagateCommit`) should be able to accept multiple copies of their inputs. This includes, `create-branch`, `start-commit`, `put/cp/delete-file`, `delete-commit`... maybe others?

**Comment by gabrielgrant** [comment]
Instead of having each command take multiple operations, what about just allowing users to wrap multiple commands in a transaction to be applied atomically?

Not 100% sure this is a good idea, but it feels like this could kind of be a nice evolution of manually calling `start-`/`finish-commit`. The semantics you'd want probably aren't identical (for instance transactions, more than commits, should probably user-scoped), but there's certainly some similarity in the idea of creating a set of operations to be applied atomically.

One thing I'm not totally sure of is how this would interact with provenance/branch-updating/commit propagation. My thinking is that new commits within a transaction would actually get created immediately so that we have somewhere to store the data, but the relevant branch pointer wouldn't get moved until calling `commit`

```
$ pc create-repo input1
$ pc create-repo input2
$ pc create-pipeline -f my-pipeline-that-takes-input1-and-input2.json
$ pc begin-transaction
76296847-ba29-4a29-b8ea-27ffaa687b12
$ pc put-file input1 master some-file 
added to transaction 76296847-ba29-4a29-b8ea-27ffaa687b12
$ pc put-file input2 master some-other-file
added to transaction 76296847-ba29-4a29-b8ea-27ffaa687b12
$ pc put-file input1 master some-third-file
added to transaction 76296847-ba29-4a29-b8ea-27ffaa687b12
$ pc commit-transaction
committing...
committed successfully
$ pc list-job --pipeline my-pipeline
# only one job has been created (i think?)
```

**Comment by gabrielgrant** [comment]
Specific request for atomic updates: https://github.com/pachyderm/pachyderm/issues/3287#issuecomment-454944627

**Comment by JoeyZwicker** [comment]
Atomic updates would also be valuable in the use case described here: https://github.com/pachyderm/pachyderm/issues/2296#issuecomment-468499268

@jdoliner certainly not urgent, but what are your thoughts for putting this into an upcoming milestone?

**Comment by Tryneus** [comment]
# Multi-update pachctl commands proposal

This is mostly a strawman as I don't have a ton of context here, so I'm not overly attached to any one approach here.  Let me know what you think and I'll get started.

## `pachctl` changes

Adds the following new commands:
* `pachctl create-branches`
* `pachctl start-commits`
* `pachctl delete-commits`
* `pachctl put-files`
* `pachctl copy-files`
* `pachctl delete-files`
* `pachctl create-pipelines`
* `pachctl start-pipelines`
* `pachctl stop-pipelines`

These commands will take a set of operations

### Complications
* `start-commit <repo> [<branch>] [<flags>]` - `branch` is optional meaning we can't just chain these together, there's no way to tell if an identifier is a branch name or a repo name
  * proposal: change `<repo> <branch>` or `<repo> <commit>` to `<repo>/<branch>` `<repo>/<commit>` ~everywhere
    * this is probably a compat issue with existing user scripting
* Should `create-pipelines` accept a JSON array, or multiple `-f` directives?
  * If `-f -` is passed we would almost certainly need to accept a JSON array, and possibly newline-delimited JSON objects
* Because the parameters to these commands are all positional, it is not visually distinctive where one operation ends and the next one begins
  * This is alleviated slightly by using `<repo>/<branch>` notation, but not fully
  * `copy-files` is especially egregious because it uses 6 positional parameters per operation
  * Supporting transactions as in @gabrielgrant's suggestion would make this problem go away, but we would need a more general solution

## proto changes
Add the following RPCs.

### PFS
```proto
rpc StartCommits(stream StartCommitRequest) returns (stream Commit) {}
rpc DeleteCommits(stream DeleteCommitRequest) returns (google.protobuf.Empty) {}
rpc CreateBranches(stream CreateBranchRequest) returns (google.protobuf.Empty) {}
rpc CopyFiles(stream CopyFileRequest) returns (google.protobuf.Empty) {}
rpc DeleteFiles(stream DeleteFileRequest) returns (google.protobuf.Empty) {}
```

### PPS
```proto
rpc CreatePipelines(stream CreatePipelineRequest) returns (google.protobuf.Empty) {}
rpc StartPipelines(stream StartPipelineRequest) returns (google.protobuf.Empty) {}
rpc StopPipelines(stream StopPipelineRequest) returns (google.protobuf.Empty) {}
```

## `driver` changes
Provide an abstraction for STM transactions and `driver.propagateCommit` so that we can perform multiple operations within one atomic transaction.  Hoist the STM functionality out of driver functions such as `createBranch` and `deleteCommit` and add a helper driver function to make this composable.

## `apiServer` changes
Add handlers for the above RPCs and modify existing handlers to adhere to the new `driver` interface.

## Alternatives
@jdoliner mentioned that `create-branch` is the command which could most use this functionality.  For now, we could just add the `create-branches` RPC and implement a more general solution in the future to save on an unnecessary middle step (see 'Future work' below).

## Future work
I believe we should end up in a world where a client can send an arbitrary set of modification requests that will be executed transactionally through STM.  Making this available through `pachctl` (as @gabrielgrant suggests above) is a little more complicated, but it shouldn't be too hard to support this through client libraries.  The `driver` changes above should be enough of a framework to make it simple to add such an RPC to the `apiServer`.

Consider deprecating the existing RPCs this proposal mirrors, although I'm not sure what the policy is for compatibility and how long we should support old clients.

**Comment by jdoliner** [comment]
This is a nice complete proposal. In terms of value I'd rank the various commands as follows:

Crucial:
- `create-branches`
This is by far the most valuable one because with this you can technically do all the other workflows, it's just not quite as nice.

Important:
- `start-commits`
- `put-files`
- `create-pipelines`
This are all heavily used and get a lot more ergonomic if they're variadic and atomic.

Nice to haves:
Everything else, these will all help in particular situations and it'd be weird not to have them be variadic and atomic eventually, but the near term value isn't huge.

# Complications

For `start-commit`, I think we could make the second argument required without breaking much, leaving the commit/branch field empty is very rarely used, it allows you to make a commit that's not on any branch, which isn't very useful. We could allow people to get the same behavior by passing an explicit empty string as the second argument, ie `pachctl start-commit <repo> ""`. That way you could in theory atomically open multiple dangling commits, I can't think of why anyone would want to do it but I think we could feel pretty comfortable compatibility wise if we did.

`create-pipeline` actually already accepts multiple json documents in the same request, it doesn't require them to be newline delimited, but they can be, so for that one we'll probably want to introduce `create-pipelines` as an alias.

We might want to even go a step further on standardizing and say that we represent paths as: `repo@branch/path/to/file` and then you'd have some prefix of that pattern for commands based on what type they take. So for repo commands you'd still just have `repo` for commit / branch commands you'd have `repo@commit` and for files you'd have the full thing. This will be a compat issue so we should probably think about this a bit longer and do the breaking pachctl changes in one fell swoop, and probably have an interim period where you can still get the old api with an env-var.

# Alternatives

I think it'd probably be a good idea to start with adding `create-branches` since that's the highest leverage one and it'll be a good test case to see how the functionality works and try out the ideas here.

**Comment by Tryneus** [comment]
One nit - doesn't matter for this current issue as it stands, but:

> `repo@branch/path/to/file`
Unfortunately branches can contain `/` characters, so this is a little ambiguous - repos cannot, however.  I'm totally on-board with a solution of this shape, though.

Another option: `repo@branch:path/to/file`

**Comment by jdoliner** [comment]
I think I like your syntax better, although I think we should probably outlaw `/`s in branch names.

**Comment by Tryneus** [comment]
This was implemented in `v1.9.0` very similarly to @gabrielgrant's proposal (disregard my proposal).  The docs about the implementation can be found in the [transaction README.md](/pachyderm/pachyderm/blob/dd06c8d96ee62c189ba36d504bc93d306acf10c7/src/server/transaction/README.md).
